### PR TITLE
Workaround clang crash due to large array assignment

### LINKED ISSUE #62664: StressLogAnalyzer is failing to build
```
  [ 98%] Built target runtimeinfo
  [100%] Built target singlefilehost
  Killed
  make[2]: *** [tools/StressLogAnalyzer/CMakeFiles/StressLogAnalyzer.dir/build.make:89: tools/StressLogAnalyzer/CMakeFiles/StressLogAnalyzer.dir/StressLogPlugin.cpp.o] Error 137
  make[1]: *** [CMakeFiles/Makefile2:5167: tools/StressLogAnalyzer/CMakeFiles/StressLogAnalyzer.dir/all] Error 2
  make: *** [Makefile:130: all] Error 2
  ~/github/wfurt-runtime/src/coreclr
  Failed to build "CoreCLR component".
/home/furt/github/wfurt-runtime/src/coreclr/runtime.proj(59,5): error MSB3073: The command ""/home/furt/github/wfurt-runtime/src/coreclr/build-runtime.sh" -x64 -release -os Linux -pgodatapath "/home/furt/.nuget/packages/optimization.linux-x64.pgo.coreclr/1.0.0-prerelease.21577.2"" exited with code 2.

```
I just sync main and my build suddenly start failing. Turns out that 4G VM is no longer good enough to build. So I added 2G more and it still fails when clang reaches 5.5G. I can probably keeping more ram but the spike in needed is very large. If I need to do that on every VM I have that would generally kill my dev machine. 

I'm wondering if there is anything we can do
cc: @am11 @jkotas @janvorli and @cshung who recently touched this ara.

### PR DESCRIPTION
**Pull request by am11**
Clang has started to kill builds on Linux systems with 4 GB memory. This failure is deterministic and always happens when compiling StressLogAnalyzer which was recently added to default build in #61123. 

<del>

Making this optional feature disablable unblocks work on such systems.

command-line usage:
```sh
# build with stresslog feature disabled
./build.sh clr -cmakeargs "-DFEATURE_STRESSLOG_ANALYZER=OFF" # or =0
```

This allows compiling coreclr in resource-constraint environments as the one @wfurt mentioned in https://github.com/dotnet/runtime/issues/62664.

</del>

Repurposed PR as we have found a workaround: https://github.com/dotnet/runtime/pull/62675#issuecomment-991933900.

cc @cshung

Fixes #62664.

### CODE REVIEW

File: src/coreclr/clrfeatures.cmake [hunk 1]
```diff
@@ -39,3 +39,7 @@ endif()
 if (CLR_CMAKE_TARGET_OSX OR CLR_CMAKE_TARGET_MACCATALYST OR CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS)
   set(FEATURE_OBJCMARSHAL 1)
 endif()
+
+if(NOT DEFINED FEATURE_STRESSLOG AND NOT CLR_CMAKE_TARGET_ALPINE_LINUX AND (CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_ARM64))
+  option(FEATURE_STRESSLOG ON)
```
Comment (cshung) [inline src/coreclr/clrfeatures.cmake:44]: Consider naming this flag `FEATURE_STRESSLOG_ANALYZER`. StressLog support is always enabled in the product, it is the analyzer that we want to turn on or off.

**Review by cshung — APPROVED** [review]
Looks good to me

**Comment by jkotas** [comment]
Do we know which part of that file is triggering the OOM? It should not be hard to find it through binary search.

**Comment by am11** [comment]
On build workers / different cloud VMs, I noticed that clang (as old as version 6 till latest v13) is crashing with SIGKILL (137) eversince #61123 is merged. Logs show it happens when build runs this command:

```sh
# directory: artifacts/obj/coreclr/Linux.x64.Debug/tools/StressLogAnalyzer/
/usr/bin/clang++-13 -DBUILDENV_DEBUG=1 -DDEBUG -DDEBUGGING_SUPPORTED -DDISABLE_CONTRACTS -DEnC_SUPPORTED -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_BASICFREEZE -DFEATURE_CODE_VERSIONING -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_COMWRAPPERS -DFEATURE_CORECLR -DFEATURE_CORESYSTEM -DFEATURE_DBGIPC_TRANSPORT_DI -DFEATURE_DBGIPC_TRANSPORT_VM -DFEATURE_DEFAULT_INTERFACES -DFEATURE_EH_FUNCLETS -DFEATURE_EVENTSOURCE_XPLAT -DFEATURE_EVENT_TRACE -DFEATURE_HIJACK -DFEATURE_ICASTABLE -DFEATURE_INSTANTIATINGSTUB_AS_IL -DFEATURE_ISYM_READER -DFEATURE_MANAGED_ETW -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES -DFEATURE_MULTICASTSTUB_AS_IL -DFEATURE_MULTICOREJIT -DFEATURE_MULTIREG_RETURN -DFEATURE_ON_STACK_REPLACEMENT -DFEATURE_PAL_ANSI -DFEATURE_PERFMAP -DFEATURE_PERFTRACING -DFEATURE_PGO -DFEATURE_PORTABLE_SHUFFLE_THUNKS -DFEATURE_PROFAPI_ATTACH_DETACH -DFEATURE_READYTORUN -DFEATURE_REJIT -DFEATURE_REMOTE_PROC_MEM -DFEATURE_STANDALONE_GC -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_TIERED_COMPILATION -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DPROFILING_SUPPORTED -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_LINUX -DTARGET_UNIX -DUNICODE -DUNIX_AMD64_ABI -DUNIX_AMD64_ABI_ITF -DURTBLDENV_FRIENDLY=Debug -D_BLD_CLR -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64 -D_SECURE_SCL=0 -D_UNICODE -I/s/runtime/src/native -I/s/runtime/src/coreclr/pal/prebuilt/inc -I/s/runtime/artifacts/obj -I/s/runtime/src/coreclr/pal/inc -I/s/runtime/src/coreclr/pal/inc/rt -I/s/runtime/src/coreclr/pal/src/safecrt -I/s/runtime/src/coreclr/inc -I/s/runtime/src/coreclr/debug/inc -I/s/runtime/src/coreclr/debug/inc/amd64 -I/s/runtime/src/coreclr/debug/inc/dump -I/s/runtime/src/coreclr/md/inc -I/s/runtime/src/coreclr/classlibnative/bcltype -I/s/runtime/src/coreclr/classlibnative/cryptography -I/s/runtime/src/coreclr/classlibnative/inc -I/s/runtime/artifacts/obj/coreclr/Linux.x64.Debug/inc -I/s/runtime/src/coreclr/hosts/inc -I/s/runtime/src/coreclr/minipal -I/s/runtime/src/coreclr/pal/inc/rt/cpp -I/s/runtime/src/coreclr/nativeresources -g -fPIE -fPIE -O0 -g -Wall -Wno-null-conversion -fno-omit-frame-pointer -fms-extensions -fwrapv -fstack-protector-strong -Werror -Wno-unused-variable -Wno-unused-value -Wno-unused-function -Wno-tautological-compare -Wno-unknown-pragmas -Wimplicit-fallthrough -Wno-invalid-offsetof -Wno-unused-but-set-variable -Wno-unknown-warning-option -ferror-limit=4096 -Wno-unused-private-field -Wno-constant-logical-operand -Wno-pragma-pack -Wno-incompatible-ms-struct -Wno-reserved-identifier -fsigned-char -fvisibility=hidden -nostdinc -std=gnu++11 -MD -MT tools/StressLogAnalyzer/CMakeFiles/StressLogAnalyzer.dir/StressLogPlugin.cpp.o -MF CMakeFiles/StressLogAnalyzer.dir/StressLogPlugin.cpp.o.d -o CMakeFiles/StressLogAnalyzer.dir/StressLogPlugin.cpp.o -c /s/runtime/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
```
it is compile step and linker is not involved in this crash. I ran that command under strace, here is the output http://sprunge.us/K6nfLD

**Comment by jkotas** [comment]
> here is the output http://sprunge.us/K6nfLD

I do not think it is possible to tell what leads to the large memory consumption from this trace. The trace just says that clang allocates a lot of memory.

I meant to perform experiment like this:
- Add `#if 0` roughly in the middle of `StressLogAnalyzer.cpp` (such that it does not introduce compile errors) and `#endif` at the end.
- Move `#if 0` up or down until you find the function or line that triggers the large memory consumption.

**Comment by am11** [comment]
Ah, I see. With the process of elimination, I've found that commenting out this line fixes the crash:
https://github.com/dotnet/runtime/blob/0e499ac03d4eacce9ce124f454d0ed4ac2cdafcd/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp#L1237
or without commenting that line, this also fixes the crash:

```diff
--- a/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
+++ b/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
@@ -561,7 +561,7 @@ static ThreadStressLogDesc s_threadStressLogDesc[MAX_THREADSTRESSLOGS];
 static int s_threadStressLogCount;
 static LONG64 s_wrappedWriteThreadCount;
 
-static const LONG MAX_MESSAGE_COUNT = 64 * 1024 * 1024;
+static const LONG MAX_MESSAGE_COUNT = 64 * 1024;
 static StressThreadAndMsg* s_threadMsgBuf;
 static volatile LONG64 s_msgCount = 0;
 static volatile LONG64 s_totalMsgCount = 0;
```
(gcc has no issue compiling this code on the same system)

**Comment by am11** [comment]
This also fixes the crash:
```diff
diff --git a/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp b/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
index 2b8b0858823..1ac8621bccf 100644
--- a/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
+++ b/src/coreclr/tools/StressLogAnalyzer/StressLogPlugin.cpp
@@ -1234,7 +1234,8 @@ int ProcessStressLog(void* baseAddress, int argc, char* argv[])
     StressLog::writing_base_address = (size_t)hdr->memoryBase;
     StressLog::reading_base_address = (size_t)baseAddress;
     s_hdr = hdr;
-    s_threadMsgBuf = new StressThreadAndMsg[MAX_MESSAGE_COUNT];
+    auto temp = new StressThreadAndMsg[MAX_MESSAGE_COUNT];
+    s_threadMsgBuf = temp;
     int threadStressLogIndex = 0;
     double latestTime = FindLatestTime(hdr);
     if (s_timeFilterStart < 0)
gitpod /workspace/template-typescri
```

If this looks ok, I'll repurpose this PR to workaround the bug using temp.

**Comment by jkotas** [comment]
Yes, repurposing this PR to workaround the bug sounds good to me.

**Comment by jkotas** [comment]
It would be nice to open a bug against clang if they do not have one already.

**Comment by wfurt** [comment]
I can confirm the fix. I shrink my VM back, sync &  everything is working again. Thanks for quick fix @am11

**Comment by am11** [comment]
> It would be nice to open a bug against clang if they do not have one already.

Reported: https://github.com/llvm/llvm-project/issues/52644. I was able to make a simple repo which always fail in GitHub Actions and a cloud VM (both running Ubuntu x64).

**Pull request merged**